rosidl_parser/rosidl_parser/definition.py

```diff
@@ -0,0 +1,343 @@
+import pathlib
+
+
+BASIC_TYPES = [
+    'short',
+    'unsigned short',
+    'long',
+    'unsigned long',
+    'long long',
+    'unsigned long long',
+    'float',
+    'double',
+    'long double',
+    'char',
+    'wchar',
+    'boolean',
+    'octet',
+    'int8',
+    'uint8',
+    'int16',
+    'uint16',
+    'int32',
+    'uint32',
+    'int64',
+    'uint64',
+]
+
+SERVICE_REQUEST_MESSAGE_SUFFIX = '_Request'
+SERVICE_RESPONSE_MESSAGE_SUFFIX = '_Response'
+
+
+class AbstractType:
+
+    __slots__ = ()
+
+    def __eq__(self, other):
+        return type(self) == type(other)
+
+
+class BaseType(AbstractType):
+
+    __slots__ = ()
+
+
+class BasicType(BaseType):
+
+    __slots__ = ('type', )
+
+    def __init__(self, typename):
+        super().__init__()
+        assert typename in BASIC_TYPES
+        self.type = typename
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.type == other.type
+
+
+class NamedType(BaseType):
+
+    __slots__ = ('name')
+
+    def __init__(self, name):
+        super().__init__()
+        self.name = name
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.name == other.name
+
+
+class NamespacedType(BaseType):
+
+    __slots__ = ('namespaces', 'name')
+
+    def __init__(self, namespaces, name):
+        super().__init__()
+        self.namespaces = namespaces
+        self.name = name
+
+    def __eq__(self, other):
+        return super().__eq__(other) and \
+            self.namespaces == other.namespaces and self.name == other.name
+
+
+class BaseString(BaseType):
+
+    __slots__ = ('maximum_size', )
+
+    def __init__(self, maximum_size=None):
+        self.maximum_size = maximum_size
+
+    def __eq__(self, other):
+        return super().__eq__(other) and \
+            self.maximum_size == other.maximum_size
+
+
+class String(BaseString):
+
+    __slots__ = ()
+
+    def __init__(self, maximum_size=None):
+        super().__init__(maximum_size=maximum_size)
+
+
+class WString(BaseString):
+
+    __slots__ = ()
+
+    def __init__(self, maximum_size=None):
+        super().__init__(maximum_size=maximum_size)
+
+
+# the following types are templated on a base type
+
+class NestedType(AbstractType):
+
+    __slots__ = ('basetype', )
+
+    def __init__(self, basetype):
+        super().__init__()
+        assert isinstance(basetype, BaseType), basetype
+        self.basetype = basetype
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.basetype == other.basetype
+
+
+class Array(NestedType):
+
+    __slots__ = ('size')
+
+    def __init__(self, basetype, size):
+        super().__init__(basetype)
+        self.size = size
+
+    def __eq__(self, other):
+        return super().__eq__(other) and self.size == other.size
+
+
+class Sequence(NestedType):
+
+    __slots__ = set()
+
+    def __init__(self, basetype):
+        super().__init__(basetype)
+
+
+class UnboundedSequence(Sequence):
```

Comment:
What's the difference between Sequence and UnboundedSequence?

Reply:
`Sequence` is the base class of `BoundedSequence` and `UnboundedSequence`.